multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,57 +93,114 @@ func NewMemberClusterSetReconciler(client client.Client,
 			}
-			r.remoteCommonArea.Stop()
-			r.remoteCommonArea = nil
-			r.clusterSetConfig = nil
-			r.clusterID = common.InvalidClusterID
-			r.clusterSetID = common.InvalidClusterSetID
 		}
+		r.commonAreaLock.Unlock()
 		return ctrl.Result{}, nil
 	}
 
 	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
 
 	// Handle create or update
-	if r.clusterSetConfig == nil {
-		r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		r.clusterSetID = common.ClusterSetID(clusterSet.Name)
-		if clusterSet.Spec.ClusterID == "" {
-			// ClusterID is a required feild, and the empty value case should only happen
-			// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
-			// Here we try to update the ClusterSet's ClusterID when it's configured in an
-			// existing ClusterClaim.
-			clusterSet.Spec.ClusterID = string(r.clusterID)
-			err = r.Update(context.TODO(), clusterSet)
-			if err != nil {
-				klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
+
+	newLeader := clusterSet.Spec.Leaders[0]
+	if r.installedLeader.clusterID == newLeader.ClusterID && r.installedLeader.serverUrl == newLeader.Server &&
+		r.installedLeader.secretName == newLeader.Secret {
+		klog.V(2).InfoS("No change for leader cluster configuration")
+		r.commonAreaLock.Unlock()
+		return ctrl.Result{}, nil
+	}
+
+	var clusterSetCreated bool
+	if r.clusterID == "" || r.clusterID == common.InvalidClusterID {
+		clusterSetCreated = true
+	}
+	r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
+	if err != nil {
+		r.commonAreaLock.Unlock()
+		return ctrl.Result{}, err
+	}
+
+	// ClusterSet deletion may fail and retry, but a ClusterSet may have been created just right before the retry.
+	// ClusterSet deletion will become an update action, so try to delete stale resources here before
+	// initilize the new ClusterSet.
+	if !clusterSetCreated {
+		if r.remoteCommonArea != nil {
+			if err := cleanUpResources(false); err != nil {
+				r.commonAreaLock.Unlock()
 				return ctrl.Result{}, err
 			}
 		}
 	}
+
+	r.clusterSetID = common.ClusterSetID(clusterSet.Name)
+	if clusterSet.Spec.ClusterID == "" {
+		// ClusterID is a required feild, and the empty value case should only happen
+		// when Antrea Multi-cluster is upgraded from an old version prior to v1.13.
+		// Here we try to update the ClusterSet's ClusterID when it's configured in an
+		// existing ClusterClaim.
+		clusterSet.Spec.ClusterID = string(r.clusterID)
+		err = r.Update(context.TODO(), clusterSet)
+		if err != nil {
+			klog.ErrorS(err, "Failed to update ClusterSet's ClusterID", "clusterset", req.NamespacedName)
+			r.commonAreaLock.Unlock()
+			return ctrl.Result{}, err
+		}
+	}
+
 	r.clusterSetConfig = clusterSet.DeepCopy()
 
 	err = r.createOrUpdateRemoteCommonArea(clusterSet)
```

Comment:
We should change this func to do only commonArea creation for the creation case, no need to stop/unset commonArea and no need to check if something changed inside the func again.

Reply:
Yeah, the change check part is moved out of it, and rename the name.